### PR TITLE
New new deploy page (WIP)

### PR DESCRIPTION
**Pull request by Calvinp**
This is a mostly working version of the react/redux new deploy page.
It's incredibly complicated, and I'm sure there are plenty of bugs I don't know about on top of the ones I do. Please take a close look.
Known bugs that I'm working on:
- [x] The implementation of default values for Select is bad. It was the only way I could get it to work, but it's not a real solution.
- [x] Sometimes an error comes back from the server as `[object object]`
- [x] Can't submit fields that are `key=value` form (the server wants a map, but gets an array of strings)
- [x] Some fields, such as CPUs, need defaults and don't have them
- [x] No client-side validation of fields (eg. making sure numeric fields are actually numbers).
- [x] No explanation to user for why they can't submit the form - this is a huge form, so if one field has a problem it'll take a long time to find that one field.
- [x] It might be better to put the form group components into their own files considering how long that file already is.
- [x] Lots of linting errors

cc @tpetr @wolfd @kwm4385

### CODE REVIEW

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 1]
```diff
@@ -0,0 +1,1303 @@
...
+import RemoveButton from '../common/RemoveButton';
+import Select from 'react-select';
+import { modifyField, clearForm } from '../../actions/form';
+import {SaveAction} from '../../actions/api/deploy';
+import OverlayTrigger from 'react-bootstrap/lib/OverlayTrigger';
+import ToolTip from 'react-bootstrap/lib/Tooltip';
+import Utils from '../../utils';
+import classNames from 'classnames';
+
+class SelectFormGroup extends Component {
```
Comment (kwm4385) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: I would definitely split this file up, starting with giving each component its own file. Also the constants can be grouped in separate files.

Comment (Calvinp) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: I'm planning on making much more use of the constants, so that makes a lot of sense.

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 2]
```diff
@@ -0,0 +1,1303 @@
...
+  getValueOrDefault(fieldId) {
+    return this.getValue(fieldId) || FLATTENED_FIELDS[fieldId].default;
+  }
+
+  isRequestDaemon() {
+    return ['SERVICE', 'WORKER'].indexOf(this.props.request.request.requestType) !== -1;
+  }
+
+  // Returns true unless the object is falsey or an empty array.
+  hasValue(value) {
```
Comment (wolfd) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: An empty array is already a falsey value.

This is essentially a no-op.

File: SingularityUI/app/components/common/RemoveButton.jsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  };
+
+  render() {
+    return (
+      <Button
+        className='remove-button'
+        id={this.props.id}
+        onClick={(event) => {
+          event.preventDefault();
+          this.props.onClick();
```
Comment (kwm4385) [inline SingularityUI/app/components/common/RemoveButton.jsx]: This can be shortened to `onClick={this.props.onClick}`

File: SingularityUI/app/views/newDeployForm.jsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import View from './view';
+
+import { Provider } from 'react-redux';
+
+import NewDeployFormPage from '../components/newDeployForm/NewDeployForm'
+
+class NewDeployFormView extends View {
```
Comment (kwm4385) [inline SingularityUI/app/views/newDeployForm.jsx]: Should be using `ReactView`

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 5]
```diff
@@ -0,0 +1,1303 @@
...
+    if (this.getValueOrDefault('executorType') === DEFAULT_EXECUTOR_TYPE) {
+      this.addFieldsToDeployObject(deployObject, FIELDS.defaultExecutor);
+    } else {
+      this.addFieldsToDeployObject(deployObject, FIELDS.customExecutor);
+    }
+    deployObject.requestId = this.props.request.request.id;
+    this.props.save({deploy: deployObject});
+  }
+
+  addThingToArrayField(fieldId, thing) {
```
Comment (wolfd) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: What's a `thing`?

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx (hunk 2 above)
Comment (andyhuang91) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: can you use _.isEmpty?

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 6]
```diff
@@ -0,0 +1,1303 @@
...
+          id = "artifacts"
+          value = {this.getValue('uris') || []}
+          onChange = {(newValue) => this.updateField('uris', newValue)}
+          placeholder="eg: http://s3.example/my-artifact"
+        />
+      </div>
+    );
+    return (
+      <div>
+        <fieldset id="default-expandable" className='expandable'>
```
Comment (wolfd) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: Run this through `eslint` (not `gulp lint`, the plugin is broken for inherited configuration for some odd reason). The className should be double quotes.

File: SingularityUI/app/components/common/formItems/MultiInput.jsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      valueClone.push(value);
+    } else {
+      valueClone[key] = value;
+    }
+    this.props.onChange(_.without(valueClone, ""));
+  };
+
+  render() {
+    const valueClone = this.props.value.slice()
+    if (valueClone.length === 0 || _.last(valueClone)) {
```
Comment (zdhickman) [inline SingularityUI/app/components/common/formItems/MultiInput.jsx]: You can use the shorthand `!valueClone.length`

File: SingularityUI/app/components/common/formItems/MultiInput.jsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      <div id={this.props.id} className={this.props.className}>
+        {
+          valueClone.map((value, key) => {
+            return (
+              <FormField
+                key = {key}
+                prop = {{
+                  value: value,
+                  updateFn: event => this.change(key, event.target.value),
+                  /* required: this.props.required, Can't mark this as form required because there will always be an extra empty field */
```
Comment (zdhickman) [inline SingularityUI/app/components/common/formItems/MultiInput.jsx]: Probably should not leave code commented out

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 9]
```diff
@@ -0,0 +1,1303 @@
...
+        <fieldset>
+          {unpauseOnSuccessfulDeploy}
+        </fieldset>
+      </div>
+    );
+
+    const errorMessage = (
+      this.props.saveApiCall.error ?
+        <p className='alert alert-danger'>
+          There was a problem saving your request: {this.props.saveApiCall.error.message}
```
Comment (kwm4385) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: You can use `boolean && <Component/>` here instead of a ternary with null

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 10]
```diff
@@ -0,0 +1,1303 @@
...
+        {hostPath}
+        {mode}
+      </div>
+    );
+  }
+
+  renderDockerVolumes() {
+    const volumes = this.getValue('volumes');
+    if (!volumes) {
+      return (<div id="docker-volumes" />);
```
Comment (kwm4385) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: ids should be avoided (this is likely an artifact from the old template/view)

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 11]
```diff
@@ -0,0 +1,1303 @@
...
+
+  componentDidMount() {
+    this.props.clearForm(FORM_ID);
+  }
+
+  updateField(fieldId, newValue) {
+    this.props.update(FORM_ID, fieldId, newValue);
+  }
+
+  getValue(fieldId) {
```
Comment (wolfd) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: Is it worth having this?

I don't see any prop validation here, so I don't know if form is a required prop, but I think it should be. Then this field can be removed.

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 12]
```diff
@@ -0,0 +1,1303 @@
...
+}
+
+const FLATTENED_FIELDS = _.extend(
+  {},
+  makeFlattenedFields(FIELDS.all),
+  makeFlattenedFields(FIELDS.customExecutor),
+  makeFlattenedFields(FIELDS.defaultExecutor)
+);
+
+class NewDeployForm extends Component {
```
Comment (wolfd) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: Add prop validation

**Comment by Calvinp** [comment]
Fixed a bunch of the bugs, now I'll get to responding to PR feedback.

File: SingularityUI/app/components/common/RemoveButton.jsx [hunk 13]
```diff
@@ -0,0 +1,19 @@
+import React, { Component, PropTypes } from 'react';
+import Button from 'react-bootstrap/lib/Button';
+import Glyphicon from 'react-bootstrap/lib/Glyphicon';
+
+export default class RemoveButton extends Component {
```
Comment (tpetr) [inline SingularityUI/app/components/common/RemoveButton.jsx]: `<Button className='remove-button'` isn't many more characters than `<RemoveButton`, but adds the burden of being yet another component to be aware of -- is it really necessary?

File: SingularityUI/app/components/newDeployForm/NewDeployForm.jsx [hunk 14]
```diff
@@ -0,0 +1,1423 @@
...
+      this.updateField(fieldId, fieldValue);
+    }
+  }
+
+  addObjectToDeployFieldPreventDefault(fieldId, deployObject, event) {
+    event.preventDefault();
+    this.addDeployObjectToArrayField(fieldId, deployObject);
+  }
+
+  removeThingFromArrayField(fieldId, key) {
```
Comment (wolfd) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: I see that you changed `addThing...` to `addDeployObjectToArrayField`, are these functions related? Are these supposed to be `deployObject`s too?

Comment (Calvinp) [inline SingularityUI/app/components/newDeployForm/NewDeployForm.jsx]: So the things that can get added are `artifact`s, `portMapping`s, and `volume`s. I don't necessarily think `deployObject` is the best name for that category of things, but I haven't been able to think of a better one. Maybe `SubField`?
